openedx/core/djangoapps/course_apps/plugins.py

```diff
@@ -0,0 +1,112 @@
+"""
+Course Apps plugin base class and plugin manager.
+"""
+from typing import Dict, Iterator, Optional
+
+from abc import ABC
+from edx_django_utils.plugins import PluginManager
+from opaque_keys.edx.keys import CourseKey
+
+
+# Stevedore extension point namespaces
+COURSE_APPS_PLUGIN_NAMESPACE = "openedx.course_app"
+
+
+class CourseApp(ABC):
+    """
+    Abstract base class for all course app plugins.
+    """
+
+    # A unique ID for the app.
+    app_id: str = ""
+    # A friendly name for the app.
+    name: str = ""
+    # A description for the app.
+    description: str = ""
+
+    @classmethod
+    def is_available(cls, course_key: CourseKey) -> bool:
```

Comment:
@xitij2000 Coming back to this PR again, I have spent some more time trying to understand how to achieve "Abstraction" in python so I am going to advise certain changes here. 

1. The methods which are "require" child classes to "must" impliment - we should decorate them with the following decorators.

```
@classmethod
@abstractmethod
def is_available(cls, course_key: CourseKey) -> bool:
    ...
```

Once we do that, we do not need to raise `raise NotImplementedError()` and replace it with `pass`. ABC class will handle this on its own. This means if any class misses implementing any of the methods this will give a runtime error when trying to instantiate/use the class.